File: doubly_linked_list/doubly_linked_list.c
#include <stdint.h>
#include <stdlib.h>
#include <stdbool.h>

#include "storage_pool.h"
#include "doubly_linked_list.h"

// Using Storage Pool

uint32_t doubly_linked_list_init(DoublyLinkedList* doubly_list, Storage_Pool* storage_pool) {
	if (doubly_list == NULL || storage_pool == NULL) return 1;

	DoublyNode* head = storage_pool_pop(storage_pool);
	if (head == NULL) return 2;	// Overflow

	doubly_list->head = head;
	doubly_list->head->left = doubly_list->head;
	doubly_list->head->right = doubly_list->head;

	doubly_list->size = 0;
	doubly_list->storage_pool = storage_pool;
	return 0;
}

uint32_t doubly_linked_list_insert_left(DoublyLinkedList* doubly_list, uint32_t info) {
	if (doubly_list == NULL) return 1;

	// 1 (P <= Avail)
	DoublyNode* P = storage_pool_pop(doubly_list->storage_pool);
	if (P == NULL) return 2;	// Overflow

	// 2
	P->info = info;

	// Increment size
	doubly_list->size++;

	// 3 Insert P at Front
	DoublyNode* X = doubly_list->head;

	P->left = X;
	P->right = X->right;
	X->right->left = P;
	X->right = P;

	return 0;
}

uint32_t doubly_linked_list_pop_left(uint32_t* info, DoublyLinkedList* doubly_list) {
	if (doubly_list == NULL || info == NULL) return 1;

	if (doubly_list->head == doubly_list->head->right) {
		return 2;  // Underflow
	}

    // Decrement size
	doubly_list->size -= 1;

    DoublyNode* X = doubly_list->head->right;
	X->left->right = X->right;
	X->right->left = X->left;

	*info = X->info;
	storage_pool_push(doubly_list->storage_pool, X);
	return 0;
}

uint32_t doubly_linked_list_insert_right(DoublyLinkedList* doubly_list, uint32_t info) {
	if (doubly_list == NULL) return 1;

	// 1 (P <= Avail)
	DoublyNode* P = storage_pool_pop(doubly_list->storage_pool);
	if (P == NULL) return 2;	// Overflow

	// 2
	P->info = info;

	// Increment size
	doubly_list->size++;

	// 3 Insert P at Rear
	DoublyNode* X = doubly_list->head->left;

	P->left = X;
	P->right = X->right;
	X->right->left = P;
	X->right = P;

	return 0;
}

uint32_t doubly_linked_list_pop_right(uint32_t* info, DoublyLinkedList* doubly_list) {
	if (doubly_list == NULL || info == NULL) return 1;

	if (doubly_list->head == doubly_list->head->right) {
		return 2;  // Underflow
	}

    // Decrement size
	doubly_list->size -= 1;

    DoublyNode* X = doubly_list->head->left;
	X->left->right = X->right;
	X->right->left = X->left;

	*info = X->info;
	storage_pool_push(doubly_list->storage_pool, X);
	return 0;
}

uint32_t doubly_linked_list_insert_node(DoublyLinkedList* doubly_list, DoublyNode* X) {
	if (doubly_list == NULL || X == NULL) return 1;

	// P <= Avail
	DoublyNode* P = storage_pool_pop(doubly_list->storage_pool);
	if (P == NULL) return 2;	// Overflow

	P->left = X;
	P->right = X->right;
	X->right->left = P;
	X->right = P;

	return 0;
}

uint32_t doubly_linked_list_delete_node(DoublyLinkedList* doubly_list, DoublyNode* X) {
	if (doubly_list == NULL || X == NULL) return 1;

	X->left->right = X->right;
	X->right->left = X->left;

	// Avail <= X
	storage_pool_push(doubly_list->storage_pool, X);

	return 0;
}

uint32_t doubly_linked_list_clear(DoublyLinkedList* list) {
	if (list->head->left != list->head) {
		// storage pool used node->left as a link
		DoublyNode* head = list->head->left;
		DoublyNode* tail = list->head->right;

		storage_pool_add_slice(list->storage_pool, head, tail);
		list->head->left = list->head;
		list->head->right = list->head;

		list->storage_pool->size += list->size;
		list->size = 0;

		return 0;
	}

	return 1;
}

uint32_t doubly_linked_list_union(DoublyLinkedList* list_a, DoublyLinkedList* list_b) {
	// Insert the list_b at right of list_a
	// Reduce the size of the list_b

	if (list_a == NULL || list_b == NULL) return 1;

	if (list_b->head->left != list_b->head) {
		DoublyNode* list_b_begin = list_b->head->right;
		DoublyNode* list_b_end = list_b->head->left;

		DoublyNode* list_a_end = list_a->head->left;

		// link end with start
		list_a_end->right = list_b_begin;
		list_b_begin->left = list_a_end;

		// link all with head
		list_b_end->right = list_a->head;
		list_a->head->left = list_b_end;

		// Empty list_b
		list_b->head->left = list_b->head;
		list_b->head->right = list_b->head;
		list_a->size += list_b->size;
		list_b->size = 0;

		return 0;
	}

	return 2;
}



//void circular_list_union(CircularList* circular_list_a, CircularList* circular_list_b) {
//	// Insert the entire circular_list_b at the right of circular circular_list_a list1
//	// Implicitly reduce size of the circular_list_b
//
//	if (circular_list_b->ptr == NULL) return;
//
//	if (circular_list_a->ptr != NULL) {
//		CircularNode* P = circular_list_a->ptr->link;
//		circular_list_a->ptr->link = circular_list_b->ptr->link;
//		circular_list_b->ptr->link = P;
//	}
//
//	circular_list_a->ptr = circular_list_b->ptr;
//	circular_list_b->ptr = NULL;
//}

File: doubly_linked_list/doubly_linked_list.py
from storage_pool import STORAGE_POOL


class DoublyNode:
        def __init__(self, info=None, left=None, right=None, link=None):
            self.info = info
            self.left = left
            self.right = right
            self.link = link

        # @property
        # def link(self):
        #     return self.left
        
        # @link.setter
        # def link(self, node):
        #     self.left = node

class DoublyLinkedList:
    """
    Node class used by the storage_pool must have 'link' and 'info' attributes.
    """
    def __init__(self, storage_pool: STORAGE_POOL):
        self.head = storage_pool.pop()
        self.head.left = self.head
        self.head.right = self.head
        self.size = 0
        self.storage_pool = storage_pool

    def insert_left(self, Y):
        # P <= Avail
        P = self.storage_pool.pop()
        P.info = Y

        # Increment size
        self.size += 1
        
        # Insert P at Front
        X = self.head
        P.left = X
        P.right = X.right
        X.right.left = P
        X.right = P
        
    def insert_right(self, Y):
        # P <= Avail
        P = self.storage_pool.pop()
        P.info = Y

        # Increment size
        self.size += 1
        
        # Insert P at Front
        X = self.head.left
        P.left = X
        P.right = X.right
        X.right.left = P
        X.right = P
        
    def pop_left(self):
        if self.head == self.head.right:
            raise Exception("List Underflow: Doubly Linked List is empty")

        # Decrement size
        self.size -= 1

        X = self.head.right
        X.left.right = X.right
        X.right.left = X.left
        
        info = X.info
        self.storage_pool.push(X)
        return info
    
    def pop_right(self):
        if self.head == self.head.right:
            raise Exception("List Underflow: Doubly Linked List is empty")

        # Decrement size
        self.size -= 1

        X = self.head.left
        X.left.right = X.right
        X.right.left = X.left
        
        info = X.info
        self.storage_pool.push(X)
        return info
    
    def insert_node(self, X):
        if X is None:
            raise Exception("Doubly Linked List insert_node method: trying to insert node next to None")

        # P <= Avail
        P = self.storage_pool.pop()

        # Increment size
        self.size += 1
        
        # Insert node
        P.left = X
        P.right = X.right
        X.right.left = P
        X.right = P

    def insert_node(self, X):
        if X is None:
            raise Exception("Doubly Linked List insert_node method: trying to insert node next to None")

        # P <= Avail
        P = self.storage_pool.pop()

        # Increment size
        self.size += 1
        
        # Insert node
        P.left = X
        P.right = X.right
        X.right.left = P
        X.right = P

    def delete_node(self, X):
        if X is None:
            raise Exception("Doubly Linked List delete_node method: trying to delete None node")
        
        if X == self.head:
            raise Exception("Doubly Linked List delete_node method: trying to delete Head node")

        # Decrement size
        self.size -= 1

        X.left.right = X.right
        X.right.left = X.left
        
        # Avail <= X
        self.storage_pool.push(X)
        
    def clear(self):
        if self.head.left == self.head:
            return

        head = self.head.right
        tail = self.head.left

        self.storage_pool.add_slice(head, tail, self.size)
        self.head.left = self.head
        self.head.right = self.head
        self.size = 0
    
    def union(self, list_b: "DoublyLinkedList"):
        if list_b is None:
            return
        
        # list_b is empty
        if list_b.head == list_b.head.left:
            return
        
        # union same list
        if self == list_b:
            return
        
        list_b_start = list_b.head.right
        list_b_end = list_b.head.left
        list_a_end = self.head.left

        # link end with start
        list_a_end.right = list_b_start
        list_b_start.left = list_a_end

        # link all with head
        list_b_end.right = self.head
        self.head.left = list_b_end

        # advance size
        self.size += list_b.size

        # empty list_b
        list_b.head.left = list_b.head
        list_b.head.right = list_b.head
        list_b.size = 0


if __name__ == "__main__":
    nodes = 128 + 1
    storage_pool = STORAGE_POOL(node_class=DoublyNode, size=nodes)
    doubly_linked_list = DoublyLinkedList(storage_pool)
    n = DoublyNode(info=1)
    n1 = DoublyNode(info=2)
    n2 = DoublyNode(info=3)

    doubly_linked_list.insert_left(n)
    doubly_linked_list.insert_left(n1)
    doubly_linked_list.insert_left(n2)
    print(doubly_linked_list.size)
    print(storage_pool.size)

    doubly_linked_list.clear()
    print(doubly_linked_list.size)
    print(storage_pool.size)
